### PR TITLE
Remove magic numbers from `WOLFSSL_ASN_TEMPLATE` code

### PR DESCRIPTION
**Pull request by julek-wolfssl**
This PR removes numbers from the `WOLFSSL_ASN_TEMPLATE` code as well as fixes misc errors in `WOLFSSL_ASN_TEMPLATE` and `USE_CERT_BUFFERS_1024` code.

### CODE REVIEW

**Review by kaleb-himes — COMMENTED** [review]
I'd like a second opinion on this, can we have @SparkiDev and/or @dgarske also review?

Let's go over the enums in source files and see if we agree it would be more optimal to place them in the header instead?

Please see comments about enums and integers, most the time it is fine to assume they are either a `const int` or `int` type but I've also seen compilers that have them as either `const unsigned int` or `unsigned int`. Because of this I always treat a computation involving an `enum` and an `int` as if it were a mixed type and I add an explicit cast for portability reasons.

**Review by dgarske — COMMENTED** [review]
In my opinion this refactor doesn't cleanup or simplify the code. The ASN template structures have fixed indexes that are well documented. Using unamed enums with not guaranteed data types seems more problematic. Deferring to @SparkiDev for his input.

File: wolfcrypt/src/asn.c [hunk 1]
```diff
@@ -2824,6 +2824,9 @@ static int SetASNIntRSA(void* n, byte* output)
 static const ASNItem intASN[] = {
     { 0, ASN_INTEGER, 0, 0, 0 }
 };
+enum {
```
[the remark this answers is not in the document]
Comment (julek-wolfssl) [reply wolfcrypt/src/asn.c:2827]: I placed the `enum`'s next to the structs that they describe. This allows the reader to know what each index refers to exactly as well as scopes the enum in the same way as the described structs. This also forces all low level asn parsing to stay in `asn.c`. No point in having the `enum` in a header and the struct in the source file.

**Comment by julek-wolfssl** [comment]
@dgarske What this refactor does is it allows modification of and preproc options in the structs. Currently adding or expanding a feature inside of the struct is a pain since this means updating all of the other indices which could prove very error prone. Additionally some of the structs are vague in places and if we ever chose to expand in those areas, this refactor would facilitate such efforts.

 This also makes the code much more readable. Even with extensive comments, its not obvious what exactly is being referred to by an index.

File: wolfcrypt/src/asn.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (kaleb-himes) [reply wolfcrypt/src/asn.c:2827]: Discussed and agree in this specific instance where the indices are well documented and code-reuse is unlikely the enums are OK inline. I do still slightly worry it provides a bad reference for future code development but if we are vigilant in our code reviews we can be sure to point out this is an exception and not the norm.

**Comment by SparkiDev** [comment]
You should have put up a small piece of work before putting up the whole thing.
Don't change the indices next to the lines in the template. You've made them longer than 80 in some cases and the enum would be enough to matchup.
I don't like the names. Either name the enum or use all caps and underscores to separate words. Don't mix.

**Comment by julek-wolfssl** [comment]
Could you give an example of what kind of name you would prefer?

**Comment by SparkiDev** [comment]
If you look at wolfssl/wolfcrypt/types.h there is a pattern there. For example:
enum wc_HashType {
    WC_HASH_TYPE_NONE ...
    ...
}

The code is internal to the C file so I don't think we need to strictly follow the usage.
enum {
    ASN_INT_IDX_INT
};
Alternatively INT_ASN_IDX_INT.

**Comment by julek-wolfssl** [comment]
I changed all the index names to uppercase versions. Words and ASN depths are separated by underscores.

>Don't change the indices next to the lines in the template.

If we start guarding sections of the structs with macro's then the index numbers won't line up with the enum values. Macro guards are already present in `certExtsASN` for example. The `WOLFSSL_CERT_EXT` guard in that struct reaches until the end of the struct but if an extension type (for example the Netscape Certificate Type with a `IGNORE_NETSCAPE_CERT_TYPE`) were to be excluded, then it would make the index after the excluded section incorrect. Is this a concern we want to address? Alternatively, the indices in the templates could be removed since the template sections are to be referenced by an enum value.

File: wolfcrypt/src/asn.c [hunk 2]
```diff
@@ -6591,27 +6635,42 @@ int wc_GetKeyOID(byte* key, word32 keySz, const byte** curveOID, word32* oidSz,
...
-                    /* PRF - default is HMAC-SHA1 */
-/*  6 */            { 2, ASN_SEQUENCE, 1, 1, 1 },
-/*  7 */                { 3, ASN_OBJECT_ID, 0, 0, 0 },
-/*  8 */                { 3, ASN_TAG_NULL, 0, 0, 1 },
-/*  9 */    { 0, ASN_SEQUENCE, 1, 1, 0 },
-               /* Encryption algorithm */
-/* 10 */       { 1, ASN_OBJECT_ID, 0, 0, 0 },
-               /* IV for CBC */
-/* 11 */       { 1, ASN_OCTET_STRING, 0, 0, 0 },
+/* PBES2PARAMSASN_IDX_KDF_SEQ                */ { 0, ASN_SEQUENCE, 1, 1, 0 },
```
Comment (SparkiDev) [inline wolfcrypt/src/asn.c]: These are long names and make really long lines.
Shorten the names to the unique part for all templates.

File: wolfcrypt/src/asn.c [hunk 3]
```diff
@@ -11610,22 +11802,24 @@ int GetName(DecodedCert* cert, int nameType, int maxIdx)
     return GetCertName(cert, full, hash, nameType, cert->source, &cert->srcIdx,
                        cert->srcIdx + length);
 #else
-    ASNGetData dataASN[certNameASN_Length];
+    DECL_ASNGETDATA(dataASN, certNameASN_Length);
```
Comment (SparkiDev) [inline wolfcrypt/src/asn.c]: Didn't bother with this one as it is too small.
Rule is that we don't want more than 100 bytes of data on stack in small stack case.

File: wolfcrypt/src/asn.c [hunk 4]
```diff
@@ -9728,30 +9893,33 @@ static int StoreRsaKey(DecodedCert* cert, const byte* source, word32* srcIdx,
     return 0;
 #endif
 #else
-    ASNGetData dataASN[rsaCertKeyASN_Length];
-    int ret;
+    int ret = 0;
```
Comment (SparkiDev) [inline wolfcrypt/src/asn.c]: Too small again.

**Review by SparkiDev — CHANGES_REQUESTED** [review]

File: wolfcrypt/src/asn.c (hunk 2 above)
Comment (julek-wolfssl) [reply wolfcrypt/src/asn.c]: Shortened to unique parts. Good suggestion!

File: wolfcrypt/src/asn.c [hunk 5]
```diff
@@ -7586,20 +7666,21 @@ int EncryptContent(byte* input, word32 inputSz, byte* out, word32* outSz,
...
         pkcs8Sz = Pkcs8Pad(NULL, inputSz, blockSz);
-        SetASN_Buffer(&dataASN[6], NULL, pkcs8Sz);
+        SetASN_Buffer(&dataASN[p8EncPbes1ASN_IDX_ENCDATA], NULL, pkcs8Sz);
 
         /* Calculate size of encoding. */
-        ret = SizeASN_Items(p8EncPbes1ASN + 1, dataASN + 1,
-                            p8EncPbes1ASN_Length - 1, &sz);
+        ret = SizeASN_Items(p8EncPbes1ASN + p8EncPbes1ASN_IDX_ENCALGO_SEQ,
+                            dataASN + p8EncPbes1ASN_IDX_ENCALGO_SEQ,
+                            p8EncPbes1ASN_Length - p8EncPbes1ASN_IDX_ENCALGO_SEQ, &sz);
```
[the remark this answers is not in the document]
Comment (julek-wolfssl) [reply wolfcrypt/src/asn.c]: I tried to address this concern in https://github.com/wolfSSL/wolfssl/pull/4582/commits/5fb2bc6f8efb3b2bc68b81d02464ff676bcae9f8 . Could you take a look if this is a satisfying solution?

File: wolfcrypt/src/asn.c (hunk 3 above)
Comment (julek-wolfssl) [reply wolfcrypt/src/asn.c]: Reverted back to the way it was.

File: wolfcrypt/src/asn.c (hunk 4 above)
Comment (julek-wolfssl) [reply wolfcrypt/src/asn.c]: Reverted back to the way it was.

File: wolfssl/wolfcrypt/asn.h [hunk 6]
```diff
@@ -606,6 +633,25 @@ WOLFSSL_LOCAL void SetASN_OID(ASNSetData *dataASN, int oid, int oidType);
...
     while (0)
 
+/* Set the node and all nodes below to not be encoded.
+ *
+ * @param [in] dataASN  Dynamic ASN data item.
+ * @param [in] node     Node which should not be encoded. Child nodes will
+ *                      also not be encoded.
+ * @param [in] dataASNLen Number of items in dataASN.
+ */
+#define SetASNItem_NoOutNode(dataASN, asn, node, dataASNLen)           \
```
Comment (SparkiDev) [inline wolfssl/wolfcrypt/asn.h:643]: Good addition!

**Review by SparkiDev — APPROVED** [review]

**Pull request merged**